File: LeetCodeSolutions/LeetCode_0222.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def countNodes(self, root: TreeNode) -> int:
        def helper(node):
            if not node:
                return 0
            return 1 + helper(node.left) + helper(node.right)

        return helper(root)

    def countNodes_2(self, root: TreeNode) -> int:
        """
        判断子树是否为完全二叉树
        对于完全二叉树, 直接计算节点个数
        """

        def helper(node):
            if not node:
                return 0
            l_height = r_height = 1
            tmp = node
            while tmp.left:
                tmp = tmp.left
                l_height += 1
            tmp = node
            while tmp.right:
                tmp = tmp.right
                r_height += 1
            if l_height == r_height:
                return 2 ** l_height - 1
            return 1 + helper(node.left) + helper(node.right)

        return helper(root)

File: LeetCodeSolutions/LeetCode_0222.cpp

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int countNodes(TreeNode* root) {
        TreeNode *tmp = root;
        int depth = 0;
        while(tmp != nullptr){
            tmp = tmp->right;
            ++depth;
        }
        if (depth == 0) return 0;
        int prev = pow(2, depth) - 1;
        int l = 0, r = pow(2, depth) - 1, ret = -1;
        while (l <= r){
            int mid = l + (r - l) / 2;
            if (valid(root, depth, mid)){
                ret = mid;
                l = mid + 1;
            }
            else r = mid - 1;
        }
        // cout << ret << ' ' << depth<< endl;
        return prev + ret + 1;
    }

    bool valid(TreeNode *root, int &depth, int &cur){
        int tmp = 1 << (depth - 1);
        TreeNode *p = root;
        while(tmp){
            if(cur & tmp){
                p = p->right;
            }
            else{
                p = p->left;
            }
            tmp >>= 1;
            if (p == nullptr) return false;
        }
        return p != nullptr;
    }
};